handlers-converters/converters/worker-policy-composite-document-converter/src/main/java/com/github/cafdataprocessing/worker/policy/composite/document/converter/CompositeDocumentWorkerTaskUpdater.java

```diff
@@ -0,0 +1,237 @@
+import com.hpe.caf.worker.document.DocumentWorkerChange.SetReferenceParams;
+import com.hpe.caf.worker.document.DocumentWorkerChange.UpdateSubdocumentParams;
+import com.hpe.caf.worker.document.DocumentWorkerChangeLogEntry;
+import com.hpe.caf.worker.document.DocumentWorkerDocument;
+import com.hpe.caf.worker.document.DocumentWorkerFailure;
+import com.hpe.caf.worker.document.DocumentWorkerFieldEncoding;
+import static com.hpe.caf.worker.document.DocumentWorkerFieldEncoding.utf8;
+import com.hpe.caf.worker.document.DocumentWorkerFieldValue;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import org.apache.commons.codec.binary.Base64;
+
+public final class CompositeDocumentWorkerTaskUpdater
+{
+
+    private CompositeDocumentWorkerTaskUpdater()
+    {
+    }
+
+    public static void updateDocument(final DocumentInterface document, final List<DocumentWorkerChangeLogEntry> changeLog)
+    {
+        changeLog.stream().filter((entry) -> (entry.changes != null)).forEachOrdered((entry) -> {
+            entry.changes.forEach((change) -> {
+                applyChange(change, document);
+            });
+        });
+    }
+
+    private static void applyChange(final DocumentWorkerChange change, final DocumentInterface document)
+    {
+        Objects.requireNonNull(change);
+        Objects.requireNonNull(document);
+
+        final SetReferenceParams setReferenceParams = change.setReference;
+        if (setReferenceParams != null) {
+            setReference(setReferenceParams.value, document);
+        }
+
+        final Map<String, List<DocumentWorkerFieldValue>> addFieldsParam = change.addFields;
+        if (addFieldsParam != null) {
+            addFields(addFieldsParam, document);
+        }
+
+        final Map<String, List<DocumentWorkerFieldValue>> setFieldsParam = change.setFields;
+        if (setFieldsParam != null) {
+            setFields(setFieldsParam, document);
+        }
+
+        final List<String> removeFieldsParam = change.removeFields;
+        if (removeFieldsParam != null) {
+            removeFields(removeFieldsParam, document);
+        }
+
+        final DocumentWorkerFailure addFailureParam = change.addFailure;
+        if (addFailureParam != null) {
+            addFailures(addFailureParam, document);
+        }
+        final List<DocumentWorkerFailure> setFailuresParam = change.setFailures;
+        if (setFailuresParam != null) {
+            setFailures(setFailuresParam, document);
+        }
+
+        final DocumentWorkerDocument addSubdocumentParam = change.addSubdocument;
+        if (addSubdocumentParam != null) {
+            addSubDocument(addSubdocumentParam, document);
+        }
+
+        final InsertSubdocumentParams insertSubdocumentParams = change.insertSubdocument;
+        if (insertSubdocumentParams != null) {
+            addSubDocument(insertSubdocumentParams.subdocument, document);
```

Comment:
what is the different in the add / insert subdocument lists with regards to functionality?  I cannot see any difference here, as they both call addSubDocument, should there be a logical difference?